proxy/coco/src/announce.rs

```diff
@@ -0,0 +1,303 @@
+//! Compute, track and announce noteworthy changes to the network.
+
+use std::collections::HashSet;
+use std::ops::Deref as _;
+
+use kv::Codec as _;
+
+use librad::keys;
+use librad::net;
+use librad::uri;
+
+use crate::error::Error;
+use crate::oid::Oid;
+use crate::peer;
+
+/// Name for the bucket used in [`kv::Store`].
+const BUCKET_NAME: &str = "announcements";
+/// Key for the single value used as cache.
+const KEY_NAME: &str = "latest";
+
+/// An update and all the required information that can be announced on the network.
+pub type Announcement = (uri::RadUrn, String, Oid);
+
+/// Announces the list of given `updates` with the [`librad::net::protocol`].
+///
+/// # Errors
+///
+/// * if the announcemnet of one of the project heads failed
+pub async fn announce(
+    protocol: net::protocol::Protocol<net::peer::PeerStorage<keys::SecretKey>, net::peer::Gossip>,
+    updates: Vec<Announcement>,
+) -> Result<(), Error> {
+    for (project_urn, head, hash) in &updates {
+        let urn = uri::RadUrn::new(
+            project_urn.id.clone(),
+            uri::Protocol::Git,
+            uri::Path::parse(head)?,
+        );
+        let have = net::peer::Gossip {
+            urn,
+            rev: Some(net::peer::Rev::Git((*hash).into())),
+            origin: Some(protocol.peer_id().clone()),
+        };
+
+        protocol.announce(have).await;
+    }
+
+    Ok(())
+}
+
+/// Builds the latset list of [`Announcement`]s for the current state of the peer.
+///
+/// # Errors
+///
+/// * if listing of the projects fails
+/// * if listing of the Refs for a project fails
+pub fn build(api: &peer::Api) -> Result<Vec<Announcement>, Error> {
+    let mut list: Vec<Announcement> = vec![];
+
+    // TODO(xla): We need to avoid the case where there is no owner yet for the peer api, there
+    // should be machinery to kick off these routines only if our app state is ready for it.
+    match api.list_projects() {
+        Err(Error::Storage(librad::git::storage::Error::Config(_err))) => Ok(list),
+        Err(err) => Err(err),
+        Ok(projects) => {
+            for project in &projects {
+                let refs = api.list_project_refs(&project.urn())?;
+
+                for (head, hash) in &refs.heads {
+                    list.push((project.urn(), head.to_string(), Oid::from(*hash.deref())));
+                }
+            }
+
+            Ok(list)
+        },
+    }
+}
+
+/// Computes the list of announcements based on the difference of the `new` and `old` state. An
+/// [`Announcement`] will be included if an entry in `new` can't be found in `old`.
+#[must_use]
+pub fn diff(old_state: &[Announcement], new_state: &[Announcement]) -> Vec<Announcement> {
```

Comment:
Then `HashSet`s can be passed here, and then `diff` is just [`difference`](https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.difference) :)